File: GeeksforGeeks/C++/Medium/Shortest Prime Path/shortest-prime-path.cpp
//{ Driver Code Starts
//Initial Template for C++
#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends
//User function Template for C++
class Solution{ 
    public:
        vector<bool>isprime;
        void seive(vector<int>&v){
            int N=10005;
            isprime.resize(N+1,true);
            isprime[0]=isprime[1]=false;
            for(int i=2;i<=N;i++){
                if(isprime[i]==true){
                    for(int j=i*i;j<=N;j+=i){
                        isprime[j]=false;
                    }
                }
            }
            for(int i=1000;i<10000;i++){
                if(isprime[i])v.push_back(i);
            }
        }
        int solve(int Num1,int Num2)
        {   
            //code here
            vector<int>v;
            seive(v);
            int n=v.size();
            vector<int>vis(10000,0);
            queue<pair<int,int>>q;
            q.push({0,Num1});
            vis[Num1]=1;
            while(!q.empty()){
             auto it=q.front();
             q.pop();
             int lvl=it.first;
             int node=it.second;
             string s=to_string(node);
             if(node==Num2)return lvl;
             for(int i=0;i<4;i++){
                 char original=s[i];
                 for(char ch='0';ch<='9';ch++){
                     if(i==0&&ch=='0')continue;
                     s[i]=ch;
                     int number=stoi(s);
                     if(number==Num2)return lvl+1;
                     if(isprime[number]&&vis[number]==0)q.push({lvl+1,number});
                     vis[number]=1;
                }
                s[i]=original;
            }
        }
        return -1;
    }
};


//{ Driver Code Starts.
signed main()
{
  int t;
  cin>>t;
  while(t--)
  {
      int Num1,Num2;
      cin>>Num1>>Num2;
      Solution obj;
      int answer=obj.solve(Num1,Num2);
      cout<<answer<<endl;
  }
}
// } Driver Code Ends

File: GeeksforGeeks/Python3/Medium/Shortest Prime Path/shortest-prime-path.py
#{ 
 # Driver Code Starts
#Initial Template for Python 3

# } Driver Code Ends
#User function Template for python3
class Solution:
    def solve (self,Num1,Num2):
        #code here
        v=[1 for _ in range(10000)]
        p = 2
        n=10000
        while (p * p <= n):
            if (v[p] == 1):
                for i in range(p * p, n, p):
                    v[i] = 0
            p += 1
        ans=0
        v[Num1]=0

        def ss(x):
            s=str(x)
            t=[]
            for i in range(4):
                for new in range(10):        
                    y=s[:i]+str(new)+s[i+1:]
                    if y[0]!='0' and y!=s and v[int(y)]==1:
                        v[int(y)]=0
                        t.append(int(y))
            return(t)

        ans=0
        q=[Num1]
        while q:
            q1=[]
            for val in q:
                new=ss(val)
                for child in new:
                    if child==Num2:
                        return(ans+1)
                    q1.append(child)
            q=q1
            ans+=1
        return(ans)


#{ 
 # Driver Code Starts.
if __name__ == '__main__': 
    t = int (input ())
    for _ in range (t):
        Num1,Num2=map(int,input().split())
        ob = Solution()
        print(ob.solve(Num1,Num2))
# } Driver Code Ends